# Hacker News
Hacker News is a site started by the startup incubator [Y Combinator]("https://www.ycombinator.com"), where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

You can find the data set [here]("https://www.kaggle.com/hacker-news/hacker-news-posts")

We're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question.
Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll compare these two types of posts to determine the following:
* Do Ask HN or Show HN receive more comments on average?
* Do posts created at a certain time receive more comments on average?








In [1]:
from csv import reader
opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn = list(read_file)

for row in hn[:5]:
    print("{} \n".format(row))
    

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'] 

['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'] 

['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'] 

['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'] 

['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'] 



In [2]:
header = hn[0]
hn = hn[1:]



for row in hn[:5]:
    print("{} \n".format(row))

['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'] 

['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'] 

['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'] 

['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'] 

['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12'] 



In [3]:
print(header)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


## Extracting Ask HN and Show HN Posts

First, we'll identify posts that begin with either Ask HN or Show HN and separate the data for those two types of posts into different lists. Separating the data makes it easier to analyze in the following steps.

In [6]:
ask_post = []
show_post =[]
other_post = []

for row in hn:
    title= row[1]
    title = title.lower()
    if title.startswith('ask hn'):
        ask_post.append(row)
    elif title.startswith('show hn'):
        show_post.append(row)
    else:
        other_post.append(row)
        
print("Ask Post {} \nShow Post {} \nOther Post {}".format(len(ask_post),len(show_post),len(other_post)))
        
    

Ask Post 1744 
Show Post 1162 
Other Post 17194


## Calculating the Average Number of Comments for Ask HN and Show HN Posts

Now that we separated ask posts and show posts into different lists, we'll calculate the average number of comments each type of post receives.

In [12]:
#Calcultae average number of comments on Ask Post 
total_ask_comments = 0
for row in ask_post:
    num_comments = int(row[4])
    total_ask_comments = total_ask_comments + num_comments
    
avg_ask_comments = total_ask_comments/len(ask_post)
print("Average Comments on Ask Post {}".format(avg_ask_comments))

#Calculate average number of comments on Show Post 
total_show_comments = 0
for row in show_post:
    num_comments = int(row[4])
    total_show_comments = total_show_comments + num_comments
avg_show_comments = total_show_comments/len(show_post)
print("Average Comments on Show Post {}".format(avg_show_comments))


Average Comments on Ask Post 14.038417431192661
Average Comments on Show Post 10.31669535283993


On average, ask posts in our sample receive approximately 14 comments, whereas show posts receive approximately 10. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

## Finding the Amount of Ask Posts and Comments by Hour Created

Next, we'll determine if we can maximize the amount of comments an ask post receives by creating it at a certain time. First, we'll find the amount of ask posts created during each hour of day, along with the number of comments those posts received. Then, we'll calculate the average amount of comments ask posts created at each hour of the day receive.

import datetime as dt 
result_list = []
for row in ask_post:
    result_list.append([row[6],int(row[4])])
count_by_hour ={}
comments_by_hours = {}

for row in result_list:
    hour = row[0]
    date_format = "%m/%d/%Y %H:%M"
    comments = int(row[1])
    hour1 = dt.datetime.strptime(hour,date_format).strftime("%H")
    if hour1 in count_by_hour:
        count_by_hour[hour1] += 1
        comments_by_hours[hour1] += comments 
    else:
        count_by_hour[hour1] = 1
        comments_by_hours[hour1] = comments 
        
comments_by_hours
        

## Calculating the Average Number of Comments for Ask HN Posts by Hour

In [25]:
# Calculate the average amount of comments `Ask HN` posts created at each hour of the day receive.
avg_by_hour = []

for row in count_by_hour:
    avg_by_hour.append([row,comments_by_hours[row]/count_by_hour[row]])
    
avg_by_hour

[['03', 7.796296296296297],
 ['12', 9.41095890410959],
 ['14', 13.233644859813085],
 ['13', 14.741176470588234],
 ['22', 6.746478873239437],
 ['15', 38.5948275862069],
 ['23', 7.985294117647059],
 ['07', 7.852941176470588],
 ['17', 11.46],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['09', 5.5777777777777775],
 ['08', 10.25],
 ['11', 11.051724137931034],
 ['10', 13.440677966101696],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['00', 8.127272727272727],
 ['05', 10.08695652173913],
 ['01', 11.383333333333333],
 ['04', 7.170212765957447],
 ['19', 10.8],
 ['06', 9.022727272727273],
 ['16', 16.796296296296298]]

## Sorting and Printing Values from a List of Lists

In [ ]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])

swap_avg_by_hour

sorted_swap = sorted(swap_avg_by_hour,reverse = True)
sorted_swap

# Sort the values and print the the 5 hours with the highest average comments.

print("Top 5 Hours for 'Ask HN' Comments\n")

for avg,hr in sorted_swap[:5]:
    print("{}, {:.2f} average comments per post".format(dt.datetime.strptime(hr,"%H").strftime("%H:%M"),avg
                                                      )
         )
    
    
    

The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

# Conclusion

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.